tap_hubspot/__init__.py

```diff
@@ -1131,6 +1220,70 @@ class Stream:
     Stream('engagements', sync_engagements, ["engagement_id"], 'lastUpdated', 'FULL_TABLE')
 ]
 
+# pylint: disable=inconsistent-return-statements
+def add_custom_streams(mode, catalog=None):
+    """
+    - In DISCOVER mode, fetch the custom schema from the API endpoint and set the schema for the custom objects.
+    - In SYNC mode, extend STREAMS for the custom objects.
+
+    Args:
+        mode (str): The mode indicating whether to DISCOVER or SYNC custom streams.
+        catalog (dict): The catalog containing stream information.
+
+    Returns:
+        None or List[str]: Returns None in DISCOVER mode and a list of custom object names in SYNC mode.
+    """
+    custom_objects_schema_url = get_url("custom_objects_schema")
+    if mode == "DISCOVER":
+        # Load Hubspot's shared schemas
+        refs = load_shared_schema_refs()
+        try:
+            for custom_object in gen_request_custom_objects("custom_objects_schema", custom_objects_schema_url, {}, 'results', "paging"):
+                stream_id = custom_object["name"]
+                STREAMS.append(Stream(stream_id, sync_custom_object_records, ['id'], 'updatedAt', 'INCREMENTAL'))
+                schema = utils.load_json(get_abs_path('schemas/shared/custom_objects.json'))
+                custom_schema = parse_custom_schema(stream_id, custom_object["properties"], is_custom_object=True)
+                schema["properties"]["properties"] = {
+                    "type": "object",
+                    "properties": custom_schema,
+                }
+
+                # Move properties to top level
+                custom_schema_top_level = {'property_{}'.format(k): v for k, v in custom_schema.items()}
+                schema['properties'].update(custom_schema_top_level)
+
+                final_schema = singer.resolve_schema_references(schema, refs)
+                custom_schema_path = get_abs_path('schemas/{}.json'.format(stream_id))
+
+                # Write data to the JSON file
+                with open(custom_schema_path, 'w') as json_file:
+                    json.dump(final_schema, json_file)
+
```

Comment:
This is a very unconventional approach for handling custom streams, 
This adds another set of disk-io operations, it may not have any significant performance degradation but we are writing these schemas from memory to disk only to be read again to memory.
can we change this implementation to something where these custom schemas might be returned to the discovery function and it processes it?

Reply:
Agreed, I have removed the disk IO operations. And stored the new custom schemas in the variable till the time of execution.